### PR TITLE
Fix not having range when calling command by using lua api

### PR DESCRIPTION
**Pull request by hoofcushion**
I also formatted this file, if you don't like, you can roll back.

### CODE REVIEW

**Comment by hoofcushion** [comment]
```lua
vim.keymap.set({"n","x"},"<leader>ae",require("fittencode.api").api.explain_code,{})
```
Problem:
 如此设置一个映射，在 visual mode 下触发，无法获取 range 参数，必须使用 ":Fitten explain_code<cr>" 才能获取
Solution:
 在 action 检测到目前处于 visual mode，先回到 normal mode，然后 nvim_buf_get_mark 获取 range ( '<, '> 要在退出 visual mode 后才会更新), 否则直接从 '<, '> 获取 range.

File: doc/tags
Comment (hoofcushion) [inline doc/tags:1]: I have no idea what this mean, i didn't change it.

File: lua/fittencode/engines/actions.lua [hunk 1]
```diff
@@ -239,11 +242,14 @@ function ActionsEngine.start_action(action, opts)
 
   Log.debug('sln: {}, eln: {}', sln, eln)
 
-  local vmode = { 'v', 'V', '<C-V>' }
   Log.debug('mode: {}', api.nvim_get_mode().mode)
-  if vim.tbl_contains(vmode, api.nvim_get_mode().mode) then
-    sln = fn.getpos("'<")[2]
-    eln = fn.getpos("'>")[2]
+  if vmode[api.nvim_get_mode().mode] then
```
Comment (hoofcushion) [inline lua/fittencode/engines/actions.lua:246]: This is main change.

**Comment by luozhiya** [comment]
Hi @hoofcushion 

Please rebase the code and keep only the main parts that have been changed

**Comment by hoofcushion** [comment]
@luozhiya Done.
But [doc/tags](https://github.com/luozhiya/fittencode.nvim/pull/57/files#diff-be1adb2dc04663ee4b63555f8ac234111c80dd4e697df2f41db999db0190bb6f) This file is changed, and I have no idea how to recovery it. Would you help to see it?

File: lua/fittencode/engines/actions.lua [hunk 2]
```diff
@@ -239,12 +241,17 @@ function ActionsEngine.start_action(action, opts)
...
```
Comment (hoofcushion) [inline lua/fittencode/engines/actions.lua:245]: I use hash map to check if the vmode is matched, and I moved the definition out of the function.

File: lua/fittencode/engines/actions.lua [hunk 3]
```diff
@@ -237,15 +239,18 @@ function ActionsEngine.start_action(action, opts)
...
+    --- Save and restore marks for visual mode
+    local m1, m2 = api.nvim_buf_get_mark(buffer, '<'), api.nvim_buf_get_mark(buffer, '>')
+    api.feedkeys('', 'nx', false)
+    sln, eln = api.nvim_buf_get_mark(buffer, '<')[1], api.nvim_buf_get_mark(buffer, '>')[1]
+    api.nvim_buf_set_mark(buffer, '<', m1[1], m1[2], {})
+    api.nvim_buf_set_mark(buffer, '>', m2[1], m2[2], {})
+  else
+    sln, eln = api.nvim_buf_get_mark(buffer, '<')[1], api.nvim_buf_get_mark(buffer, '>')[1]
   end
+  Log.debug('sln: {}, eln: {}', sln, eln)
```
Comment (hoofcushion) [inline lua/fittencode/engines/actions.lua:253]: Log.debug is moved too.

File: lua/fittencode/engines/actions.lua [hunk 4]
```diff
@@ -237,15 +239,18 @@ function ActionsEngine.start_action(action, opts)
...
-    sln = fn.getpos("'<")[2]
-    eln = fn.getpos("'>")[2]
-    Log.debug('v mode sln: {}, eln: {}', sln, eln)
+  if vmode[api.nvim_get_mode().mode] then
+    --- Save and restore marks for visual mode
+    local m1, m2 = api.nvim_buf_get_mark(buffer, '<'), api.nvim_buf_get_mark(buffer, '>')
+    api.feedkeys('', 'nx', false)
+    sln, eln = api.nvim_buf_get_mark(buffer, '<')[1], api.nvim_buf_get_mark(buffer, '>')[1]
+    api.nvim_buf_set_mark(buffer, '<', m1[1], m1[2], {})
+    api.nvim_buf_set_mark(buffer, '>', m2[1], m2[2], {})
```
Comment (luozhiya) [inline lua/fittencode/engines/actions.lua:249]: 请问一下，这里为什么需要设置回去？

感觉既然调用esc切换了模式后，那么按neovim的逻辑就是要改变mark的

**Comment by luozhiya** [comment]
> ```lua
> vim.keymap.set({"n","x"},"<leader>ae",require("fittencode.api").api.explain_code,{})
> ```
> 
> Problem: 如此设置一个映射，在 visual mode 下触发，无法获取 range 参数，必须使用 `":Fitten explain_code<cr>"` 才能获取 Solution: 在 action 检测到目前处于 visual mode，先回到 normal mode，然后 nvim_buf_get_mark 获取 range ( `'<, '>` 要在退出 visual mode 后才会更新), 否则直接从 `'<, '>` 获取 range.

@hoofcushion 

API 调用可以简化一下，可以直接在 `fittencode` 这个模块下访问
```lua
require("fittencode").explain_code
```

**Comment by luozhiya** [comment]
> @luozhiya Done. But [doc/tags](https://github.com/luozhiya/fittencode.nvim/pull/57/files#diff-be1adb2dc04663ee4b63555f8ac234111c80dd4e697df2f41db999db0190bb6f) This file is changed, and I have no idea how to recovery it. Would you help to see it? I enabled `Allow edits and access to secrets by maintainers`

GitHub 上操作好像不太方便，麻烦本地使用一下如下 git 语句

```bash
git rm --cache doc/tags
git commit --amend --no-edit
git push --force
```

File: lua/fittencode/engines/actions.lua (hunk 4 above)
Comment (hoofcushion) [reply lua/fittencode/engines/actions.lua:249]: 我的习惯是不改变，如果需要改变的话就不需要设置回去。

**Comment by hoofcushion** [comment]
好了

**Comment by luozhiya** [comment]
@hoofcushion  谢谢

**Pull request merged**